rectools/models/ranking/candidate_ranking.py

```diff
@@ -0,0 +1,560 @@
+    """
+
+    # TODO: this class can be used in pipelines directly. it will keep scores and ranks and add nothing
+    # TODO: create an inherited class that will get all features from dataset?
+
+    def _get_user_features(
+        self, users: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.User])
+
+    def _get_item_features(
+        self, items: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.Item])
+
+    def _get_user_item_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=Columns.UserItem)
+
+    def collect_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        """
+        Collect features for users-item pairs from any desired sources.
+
+        Parameters
+        ----------
+        useritem : pd.DataFrame
+            Candidates with score/rank features from first stage. Ids are either external or 1x internal
+        dataset : Dataset
+            Dataset will have either external -> 2x internal id maps to internal -> 2x internal
+        fold_info : tp.Optional[tp.Dict[str, tp.Any]]
+            Fold inofo from splitter can be used for adding time-based features
+
+        Returns
+        -------
+        pd.DataFrame
+            `useritem` dataframe enriched with features for users, items and useritem pairs
+        """
+        user_features = self._get_user_features(useritem[Columns.User].unique(), dataset, fold_info)
+        item_features = self._get_item_features(useritem[Columns.Item].unique(), dataset, fold_info)
+        useritem_features = self._get_user_item_features(useritem, dataset, fold_info)
+
+        res = (
+            useritem.merge(user_features, on=Columns.User, how="left")
+            .merge(item_features, on=Columns.Item, how="left")
+            .merge(useritem_features, on=Columns.UserItem, how="left")
+        )
+        return res
+
+
+@attr.s(auto_attribs=True)
+class NegativeSamplerBase:
+    """TODO: Documentation""" ""
+
+    def sample_negatives(self, train: pd.DataFrame) -> pd.DataFrame:
+        """TODO: Documentation""" ""
+        raise NotImplementedError()
+
+
+@attr.s(auto_attribs=True)
+class PerUserNegativeSampler(NegativeSamplerBase):
+    """TODO: Documentation""" ""
+
+    n_negatives: int = 3
+    random_state: tp.Optional[int] = None
+
+    def sample_negatives(self, train: pd.DataFrame) -> pd.DataFrame:
+        """TODO: Documentation""" ""
+        # train: user_id, item_id, scores, ranks, target(1/0)
+
+        # TODO: refactor for faster computations: avoid shuffle and apply
```

Comment:
please add a link to my comment with suggestions about it

Reply:
**done**